File: outco/curriculum/target_practice_solutions/swift/w3_d3_matrix.swift
/*
* Target Practice - Matrix
*
*  Problem 1:  Robot Paths
*
*  Prompt:   Given a matrix of zeroes, determine how many unique paths exist
*            from the top left corner to the bottom right corner
*
*  Input:    An Array of Array of Integers (matrix)
*  Output:   Integer
*
*  Example:  matrix = [[0,0,0,0],
*                      [0,0,0,0],
*                      [0,0,0,0]]
*
*            robotPaths(matrix) = 38
*
*
*            matrix = [[0,0,0],
*                      [0,0,0]]
*
*            robotPaths(matrix) = 4
*
*  Note:     From any point, you can travel in the four cardinal directions
*            (north, south, east, west). A path is valid as long as it travels
*            from the top left corner to the bottom right corner, does not go
*            off of the matrix, and does not travel back on itself
*/


func robotPaths(_ matrix: inout [[Int]]) -> Int {
  func traverse(_ x: Int, _ y: Int) -> Int {
    if(x < 0 || y < 0 || x >= matrix[0].count || y >= matrix.count) {
      return 0
    } else if (matrix[y][x] == 1) {
      return 0
    } else if (x == matrix[0].count - 1 && y == matrix.count - 1) {
      return 1
    }

    matrix[y][x] = 1
    var sum = 0
    sum += traverse(x + 1, y)
    sum += traverse(x - 1, y)
    sum += traverse(x, y + 1)
    sum += traverse(x, y - 1)
    matrix[y][x] = 0
    return sum
  }
  return traverse(0, 0)
}


/*
 *  Problem 2: Matrix Spiral
 *
 *  Prompt: Given an (MxN) matrix of integers, return an array in spiral order.
 *
 *
 *  Input:  matrix {Integer[][]}
 *  Output: {Integer}
 *
 *
 * Example:  Input:  [[ 1, 2, 3],
 *                    [ 4, 5, 6],
 *                    [ 7, 8, 9]]
 *
 *  Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]
 *
 */


// Time Complexity: O(MN)
// Auxiliary Space Complexity: O(MN)




func matrix_spiral(_ matrix: [[Int]]) -> [Int] {
  if(matrix.count == 0) {return [] }
  var yMin: Int = 0
  var xMin: Int = 0
  var yMax: Int = matrix.count - 1
  var xMax: Int = matrix[0].count - 1
  var results: [Int] = []

  while (xMin <= xMax && yMin <= yMax) {

      for i in stride(from: xMin, through: xMax, by: 1) {
          results.append(matrix[yMin][i])
      }

      yMin += 1
      for i in stride(from: yMin, through: yMax, by: 1) {
          results.append(matrix[i][xMax])
      }
      xMax -= 1
      if(yMin <= yMax){
          for i in stride(from: xMax, through: xMin, by: -1) {
              results.append(matrix[yMax][i])
          }
          yMax -= 1
      }
      if(xMin <= xMax){
          for i in stride(from: yMax, through: yMin, by: -1) {
              results.append(matrix[i][xMin])
          }
          xMin += 1
      }
  }
  return results

}



////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////


// custom assert function to handle tests
// input: count {Array} - keeps track out how many tests pass and how many total
//        in the form of a two item array i.e., [0, 0]
// input: name {String} - describes the test
// input: test {Function} - performs a set of operations and returns a boolean
//        indicating if test passed
// output: {Void}
func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
  if count.count != 2 {
    count = [0, 0]
  } else {
    count[1] = count[1] + 1
  }

  var pass: String = "false"

  if test() {
    pass = " true"
    count[0] = count[0] + 1
  }
  print(count[1], ")   ", pass, ":", name)
}

print("Robot Paths Tests")
var testCount: [Int] = [0, 0]

assert(&testCount, "should work on first example input", {
  var test: [[Int]] = [[0,0,0,0],
                       [0,0,0,0],
                       [0,0,0,0]]
  let example: Int = robotPaths(&test)
  return example == 38
})

assert(&testCount, "should work on second example input", {
  var test: [[Int]] = [[0,0,0],
                       [0,0,0]]
  let example: Int = robotPaths(&test)
  return example == 4
})

assert(&testCount, "should work on single-element input", {
  var test: [[Int]] = [[0]]
  let example: Int = robotPaths(&test)
  return example == 1
})

assert(&testCount, "should work on single-row input", {
  var test: [[Int]] = [[0,0,0,0,0,0]]
  let example: Int = robotPaths(&test)
  return example == 1
})

assert(&testCount, "should work on single-column input", {
  var test: [[Int]] = [[0],
                       [0],
                       [0],
                       [0],
                       [0]]
  let example: Int = robotPaths(&test)
  return example == 1
})

assert(&testCount, "should work on larger 5x6 matrix input", {
  var test: [[Int]] = [[0,0,0,0,0,0],
                       [0,0,0,0,0,0],
                       [0,0,0,0,0,0],
                       [0,0,0,0,0,0],
                       [0,0,0,0,0,0],]
  let example: Int = robotPaths(&test)
  return example == 79384
})

print("PASSED: ", testCount[0], "/", testCount[1], "\n\n")


print("Matrix Spiral Tests")
testCount = [0, 0]


assert(&testCount, "should work for empty matrix input", {
  let test: [[Int]] = [[]]
  let example: [Int] = matrix_spiral(test)
  return example == []
})

assert(&testCount, "should work for single element matrix input", {
  let test: [[Int]] = [[1]]
  let example: [Int] = matrix_spiral(test)
  return example == [1]
})

assert(&testCount, "should work for single column matrix input", {
  let test: [[Int]] = [[1],
                       [2],
                       [3],
                       [4]]
  let example: [Int] = matrix_spiral(test)
  return example == [1, 2, 3, 4]
})

assert(&testCount, "should work for square matrix input", {
  let test: [[Int]] = [[1, 2], [4, 3]]
  let example: [Int] = matrix_spiral(test)
  return example == [1, 2, 3, 4]
})

assert(&testCount, "should work for single row matrix input", {
  let test: [[Int]] = [[1, 2, 3, 4]]
  let example: [Int] = matrix_spiral(test)
  return example == [1, 2, 3, 4]
})

assert(&testCount, "should work for 3 x 5 matrix input", {
  let test: [[Int]] = [[ 1,  2,  3],
                       [ 4,  5,  6],
                       [ 7,  8,  9],
                       [10, 11, 12],
                       [13, 14, 15]]
  let example: [Int] = matrix_spiral(test)
  return example == [1, 2, 3, 6, 9, 12, 15, 14, 13, 10, 7, 4, 5, 8, 11]
})

assert(&testCount, "should work for 5 x 3 matrix input", {
  let test: [[Int]] = [[  1,  2,  3,  4, 5],
                       [ 12, 13, 14, 15, 6],
                       [ 11, 10,  9,  8, 7]]
  let example: [Int] = matrix_spiral(test)
  return example == [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
})

print("PASSED: ", testCount[0], "/", testCount[1], "\n\n")

File: outco/curriculum/target_practice_solutions/rb/w3_d3_matrix.rb
#  Target Practice - Matrix
#
#  Problem 1:  Robot Paths
#
#  Prompt:   Given a matrix of zeroes, determine how many unique paths exist
#            from the top left corner to the bottom right corner
#
#  Input:    An Array of Array of Integers (matrix)
#  Output:   Integer
#
#  Example:  matrix = [[0,0,0,0],
#                      [0,0,0,0],
#                      [0,0,0,0]]
#
#            robotPaths(matrix) = 38
#
#
#            matrix = [[0,0,0],
#                      [0,0,0]]
#
#            robotPaths(matrix) = 4
#
#  Note:     From any point, you can travel in the four cardinal directions
#            (north, south, east, west). A path is valid as long as it travels
#            from the top left corner to the bottom right corner, does not go
#            off of the matrix, and does not travel back on itself

def robot_paths(matrix)
  traverse = lambda do |x, y|
    if x < 0 || y < 0 || x >= matrix[0].length || y >= matrix.length
      return 0
    elsif matrix[y][x] == 1
      return 0
    elsif x == matrix[0].length - 1 && y == matrix.length - 1
      return 1
    end

    matrix[y][x] = 1
    sum = 0
    sum += traverse.call(x+1, y)
    sum += traverse.call(x-1, y)
    sum += traverse.call(x, y+1)
    sum += traverse.call(x, y-1)
    matrix[y][x] = 0
    return sum
  end
  return traverse.call(0, 0)
end


# Problem # 2: Matrix spiral
#
# Given an (MxN) matrix of integers, return an array in spiral order.
#
# Input:  matrix {Integer[][]}
# Output: {Integer}
#
#
# Example:
# Input:  [[ 1, 2, 3],
#          [ 4, 5, 6],
#          [ 7, 8, 9]]
# Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]
#
#

# Time Complexity: O(MN)
# Auxiliary Space Complexity: O(MN)



def matrix_spiral(input)
  output = []
  row_min, col_min = 0, 0
  row_max, col_max = input.length - 1, input[0].length - 1
  while row_min <= row_max && col_min <= col_max

    for col in col_min ... col_max + 1 do
      output.push(input[row_min][col])
    end
    row_min += 1

    for row in row_min ... row_max + 1 do
      output.push(input[row][col_max])
    end
    col_max -= 1

    if row_min <= row_max
      col_max.downto(col_min) do |col|
        output.push(input[row_max][col])
      end
      row_max -= 1
    end

    if col_min <= col_max
      row_max.downto(row_min) do |row|
        output.push(input[row][col_min])
      end
      col_min += 1
    end

  end
  output
end



############################################################
###############  DO NOT TOUCH TEST BELOW!!!  ###############
############################################################

# custom expect function to handle tests
# List count : keeps track out how many tests pass and how many total
#   in the form of a two item array i.e., [0, 0]
# String name : describes the test
# Function test : performs a set of operations and returns a boolean
#   indicating if test passed
def expect(count, name, test)
  count[1] = count[1] + 1

  result = 'false'
  err_msg = nil
  begin
    if test()
      result = ' true'
      count[0] = count[0] + 1
    end
  rescue NameError => e
    MyModule::Logger.error("Error loading the deployer #{class_name}. This deployer it's not installed!")
  rescue StandardError => err
    err_msg = err.message
  end

  puts'  ' + (count[1]).to_s + ')   ' + result + ' : ' + name
  if err_msg != nil
    puts'       ' + err_msg + '\n'
  end
end

# code for capturing puts output
require 'stringio'
require 'ostruct'

class Capture

  def self.capture &block

    # redirect output to StringIO objects
    stdout = StringIO.new
    $stdout = stdout

    result = block.call

    # restore normal output
    $stdout = STDOUT

    OpenStruct.new result: result, stdout: stdout.string
  end
end

# code for checking if lists are equal
def arrays_equal(arr1, arr2)
  if arr1.length != arr2.length
    return false
  end
  for i in 0..arr1.length-1
    if arr1[i] != arr2[i]
      return false
    end
  end
  return true
end

# code for checking if array is sorted (linear runtime)
def is_sorted(input)
  if (input.length < 2)
    return true
  end
  for i in 1..input.length - 1
    if (input[i-1] > input[i])
      return false
    end
  end
  return true
end


puts 'Robot Path Tests'
test_count = [0, 0]

def test()
  input = [[0,0,0,0],
          [0,0,0,0],
          [0,0,0,0]]
  example = robot_paths(input)
  return example == 38
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should work on first example input', test)

def test()
  input = [[0,0,0],
          [0,0,0]]
  example = robot_paths(input)
  return example == 4
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should work on second example input', test)

def test()
  input = [[0]]
  example = robot_paths(input)
  return example == 1
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should work on single-element input', test)

def test()
  input = [[0,0,0,0,0,0]]
  example = robot_paths(input)
  return example == 1
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should work on single-row input', test)

def test()
  input = [[0],
           [0],
           [0],
           [0],
           [0]]
  example = robot_paths(input)
  return example == 1
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should work on single-column input', test)

def test()
  input = [[0,0,0,0,0,0],
           [0,0,0,0,0,0],
           [0,0,0,0,0,0],
           [0,0,0,0,0,0],
           [0,0,0,0,0,0]]
  example = robot_paths(input)
  return example == 79384
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should work on a 5 x 6 matrix input', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"



puts "\nMatrix Spiral"
test_count = [0, 0]

def test()
  matrix = [[]]
  example = matrix_spiral(matrix)
  return example == []
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on empty matrix input", test)

def test()
  matrix = [[1]]
  example = matrix_spiral(matrix)
  return example == [1]
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on single element matrix input", test)


def test()
  matrix =  [[1],
             [2],
             [3],
             [4],
             [5]]
  example = matrix_spiral(matrix)
  return example == [1, 2, 3, 4, 5]
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on single column matrix input", test)

def test()
  matrix = [[1, 2],
            [4, 3]]
  example = matrix_spiral(matrix)
  return example == [1, 2, 3, 4]
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on square matrix input", test)


def test()
  matrix = [[1, 2, 3, 4]]
  example = matrix_spiral(matrix)
  return example == [1, 2, 3, 4]
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on single row matrix input", test)


def test()
  matrix = [[ 1,  2,  3],
            [ 4,  5,  6],
            [ 7,  8,  9],
            [10, 11, 12],
            [13, 14, 15]]
  example = matrix_spiral(matrix)
  return example == [1, 2, 3, 6, 9, 12, 15, 14, 13, 10, 7, 4, 5, 8, 11]
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on 3 x 5 matrix input", test)


def test()
  matrix =  [[  1,  2,  3,  4, 5],
             [ 12, 13, 14, 15, 6],
             [ 11, 10,  9,  8, 7]]
  example = matrix_spiral(matrix)
  return example == [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
rescue StandardError => err
  puts err.message
end
expect(test_count, "should work on 5 x 3 matrix input", test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"
